### PR TITLE
Implement RxCopy

### PR DESCRIPTION
**Pull request by Sammers21**
This Pull Request provides an implementation for RxCopy class

Can be merged after #164

### CODE REVIEW

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 4@dc9a75c2-99a9-40a1-bc27-ddaf8f08c10c, thread: PQ-CT2E6TK9B
com.zerocracy.farm.reactive.StkRuntime[76] java.lang.IllegalStateException: org.codehaus.groovy.runtime.InvokerInvocationException in com.zerocracy.stk.pm.staff.elections.elect_performer
org.codehaus.groovy.reflection.CachedMethod[105] org.codehaus.groovy.runtime.InvokerInvocationException: java.lang.OutOfMemoryError: unable to create new native thread
java.lang.Thread[-2] java.lang.OutOfMemoryError: unable to create new native thread
```

[0.53.12](https://github.com/zerocracy/farm): CID: [955700a3-e474-43c1-96f7-7f81776f3876](https://www.0crat.com/footprint/CT2E6TK9B/955700a3-e474-43c1-96f7-7f81776f3876), Type: "Elect performer"

**Comment by 0crat** [comment]
This pull request [#165](https://github.com/artipie/asto/issues/165) is assigned to @hdouss[/z](https://www.0crat.com/u/hdouss), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/3564db94-eae3-453c-b09e-8e57dc2ba832); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by g4s8** [comment]
@Sammers21 I don't think we really need draft PRs, it's not really informative until it's ready for merge.

**Comment by Sammers21** [comment]
@g4s8, those Pull Request includes changes from other non merged pull requests. If I would allow reviewers to review them, they would do the same work twice.

**Comment by HDouss** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/asto/issues/165#issuecomment-619969795))

@HDouss The user @hdouss[/z](https://www.0crat.com/u/hdouss) resigned from [#165](https://github.com/artipie/asto/issues/165), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/asto/issues/165#issuecomment-619969795))

@HDouss Job refused in 2 hours - no penalty, see [§6](http://www.zerocracy.com/policy.html#6)

**Comment by Sammers21** [comment]
@g4s8 , @HDouss, the pull request is ready to reviewed

File: src/main/java/com/artipie/asto/rx/RxCopy.java [hunk 1]
```diff
@@ -67,6 +66,15 @@ public RxCopy(final RxStorage from, final List<Key> keys) {
      * @return The completion signal.
      */
     public Completable copy(final RxStorage to) {
-        return Completable.error(new IllegalStateException("not implemented"));
+        return Completable.merge(
```
Comment (g4s8) [inline src/main/java/com/artipie/asto/rx/RxCopy.java:69]: @Sammers21 I think we need to handle the parallelness of copy, starting all downloads simultaneously may cause huge IO loads or OOM errors. E.g. I'm going to use this class in `rpm-adapter` which deals with thousands of storage objects, starting 5000 downloads at the same time is not a good idea. Maybe you can use IO scheduler here.

**Comment by HDouss** [comment]
@Sammers21 I resigned from this one, I made my reviews in #164

File: src/main/java/com/artipie/asto/rx/RxCopy.java [hunk 2]
```diff
@@ -67,6 +66,15 @@ public RxCopy(final RxStorage from, final List<Key> keys) {
...
+            this.keys.stream()
```
Comment (g4s8) [inline src/main/java/com/artipie/asto/rx/RxCopy.java:70]: @Sammers21 I don't like this mix of streams API and RxJava. I'd change it to:
```java
Observable.fromIterable(keys).flatMapCompletable(
  key -> from.value(key).flatMapComplatable(content -> to.save(key, content))
);
```

**Comment by g4s8** [comment]
@0crat assign me

File: src/main/java/com/artipie/asto/rx/RxCopy.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/com/artipie/asto/rx/RxCopy.java:70]: What is wrong with the "mix"?

**Comment by Sammers21** [comment]
@g4s8, review the pull request again please

File: src/main/java/com/artipie/asto/rx/RxCopy.java [hunk 3]
```diff
@@ -40,6 +40,11 @@
 @SuppressWarnings({"PMD.UnusedPrivateField", "PMD.SingularField"})
 public class RxCopy {
 
+    /**
+     * The default parallelism level.
+     */
+    private static final Integer DEFLT_PARALLELISM = 8;
```
Comment (g4s8) [inline src/main/java/com/artipie/asto/rx/RxCopy.java:46]: @Sammers21 I would be more correct to use `Runtime.getRuntime().availableProcessors()` (like in `ParallelFlowable`)

File: src/main/java/com/artipie/asto/rx/RxCopy.java [hunk 4]
```diff
@@ -66,15 +87,16 @@ public RxCopy(final RxStorage from, final List<Key> keys) {
...
-        return Completable.merge(
-            this.keys.stream()
-                .map(
-                    key -> to.save(
-                        key,
-                        new Content.From(this.from.value(key).flatMapPublisher(cnt -> cnt))
+        return Completable.concat(Flowable.fromIterable(this.keys)
+            .map(
+                key ->
+                    Completable.defer(
```
Comment (g4s8) [inline src/main/java/com/artipie/asto/rx/RxCopy.java:93]: @Sammers21 why to use extra newline here? In other places we're starting expressions and statements on the same line as lambda params

**Comment by 0crat** [comment]
@g4s8[/z](https://www.0crat.com/u/g4s8) resigned from [#165](https://github.com/artipie/asto/issues/165), since the job is not in scope anymore

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 4@48fbdee7-d1a9-4655-9670-2586da96f316, thread: PQ-CT2E6TK9B
com.zerocracy.claims.ClaimIn[214] java.lang.IllegalArgumentException: Parameter "age" not found in "Order was canceled" among: [reason, priority, job, cause, _parent_sig, flow, voluntarily, login]
```

[0.53.14](https://github.com/zerocracy/farm): CID: [e90626da-fcee-4b03-a36b-afc1fe66b36a](https://www.0crat.com/footprint/CT2E6TK9B/e90626da-fcee-4b03-a36b-afc1fe66b36a), Type: "Order was canceled"

**Pull request merged**